File: Python/0139-word-break.py
# Time:  O(n * l), l is the max length of the words.
# Space: O(n)

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        n = len(s)
        max_len = 0
        for string in wordDict:
            max_len = max(max_len, len(string))

        can_break = [False for _ in range(n + 1)]
        can_break[0] = True

        for i in range(1, n + 1):
            for l in range(1, min(i, max_len) + 1):
                if can_break[i-l] and s[i-l:i] in wordDict:
                    can_break[i] = True
                    break

        return can_break[-1]

# Notes:
# It should have been s[i-l+1:i+1]. But we changed to s[i-l:i] as our
# string array starts from index 0

File: C++/0139-word-break.cpp
// Tags: DP Amazon
// Time:  O(n * l), l is the max length of the words, (if unordered_set is used)
// Space: O(n)
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        const int n = s.length();
        size_t max_len = 0;
        // unordered_set<string> word_set(wordDict.begin(), wordDict.end());
        for (const auto &str: wordDict)
            max_len = max(max_len, str.length());

        vector<bool> canBreak(n+1, false);
        canBreak[0] = true;

        for (int i = 1; i <= n; ++i)
            for (int l = 1; l <= max_len && i - l >= 0; ++l) {
                // if (canBreak[i - l] && word_set.count(s.substr(i - l + 1 - 1, l)))
                // if unordered_set is used
                if (canBreak[i - l] && contain(wordDict, s.substr(i - l, l))) {
                    canBreak[i] = true;
                    break;
                }
            }
        return canBreak[n];
    }

    bool contain(vector<string>& wordDict, string s){
        for(const auto &word: wordDict){
            if(s == word) return true;
        }
        return false;
    }
};
